Forge.TreeWalker/src/TreeNodeContext.cs

```diff
@@ -0,0 +1,105 @@
+        /// </summary>
+        public dynamic Properties { get; private set; }
+
+        /// <summary>
+        /// The dynamic user-defined context object that is able to be referenced when evaluating schema expressions or performing actions.
+        /// </summary>
+        public object UserContext { get; private set; }
+
+        /// <summary>
+        /// The cancellation token.
+        /// </summary>
+        public CancellationToken Token { get; private set; }
+
+        /// <summary>
+        /// The name of the ForgeTree in the JsonSchema.
+        /// </summary>
+        public string TreeName { get; private set; }
+
+        /// <summary>
+        /// The unique identifier for the root/parent tree walking session.
+        /// </summary>
+        public Guid RootSessionId { get; private set; }
+
+        /// <summary>
+        /// Indicates whether the tree walker should skip all actions defined in this node, and proceed with ChildSelector.
+        /// When not null or whitespace, the tree walker will skip all actions defined in the current tree node, and proceed to AfterVisitNode then ChildSelector.
+        /// Update this property inside BeforeVisitNode if you wish to use this feature for the current tree node.
+        /// The string context is available to check in the current TreeNode's ChildSelector via Session.GetCurrentNodeSkipActionContext().
+        /// </summary>
+        /// <remarks>Allow Public set for testing purpose.</remarks>
+        public string CurrentNodeSkipActionContext { get; set; }
+
+        /// <summary>
+        /// Instantiates an TreeNodeContext object.
+        /// </summary>
+        /// <param name="sessionId">The unique identifier for this tree walking session.</param>
+        /// <param name="treeNodeKey">The TreeNode's key.</param>
+        /// <param name="properties">The properties of this Tree node..</param>
+        /// <param name="userContext">Optional, the user context for this Tree node.</param>
+        /// <param name="token">The cancellation token.</param>
+        /// <param name="treeName">The name of the ForgeTree in the JsonSchema.</param>
+        /// <param name="rootSessionId">The unique identifier for the root/parent tree walking session.</param>
+        /// <param name="currentNodeSkipActionContext">
+        /// The string context if the actions in this TreeNodes are skipped. It is null when no skip. When it is not empty, proceed with ChildSelector.
+        /// </param>
+        public TreeNodeContext(
+            Guid sessionId,
+            string treeNodeKey,
+            dynamic properties,
+            object userContext,
+            CancellationToken token,
+            string treeName,
+            Guid rootSessionId,
+            string currentNodeSkipActionContext)
+        {
+            // userContext is an optional parameter for TreeNode, so we don't throw ArgumentNullException when null.
```

Comment:
>// userContext is an optional parameter for TreeNode, so we don't throw ArgumentNullException when null. [](start = 12, length = 104)

Please remove this comment. UserContext being an optional parameter is established in TreeWalkerParameters as well as on the wiki. We don't need to remind folks in more places of this fact.

Reply:
sure, will remove

---
In reply to: [505812830](https://github.com/microsoft/Forge/pull/32#discussion_r505812830) [](ancestors = 505812830)